src/monmove.cpp

```diff
@@ -89,6 +154,41 @@ bool monster::can_move_to( const tripoint &p ) const
         }
     }
 
+    if (has_flag( MF_INTELLIGENT_PATHS)) {
+        // Don't enter sharp terrain unless tiny, or attacking
+        if( g->m.has_flag( "SHARP", p ) && !( attitude( &( g->u ) ) == MATT_ATTACK ||
+                                              type->size == MS_TINY || has_flag( MF_FLIES ) ) ) {
+            return false;
+        }
+
+        // Don't enter open pits ever unless tiny, can fly or climb well
+        if( !( type->size == MS_TINY || can_climb ) &&
+            ( g->m.ter( p ) == t_pit || g->m.ter( p ) == t_pit_spiked || g->m.ter( p ) == t_pit_glass ) ) {
+            return false;
+        }
+
+        // Don't enter lava ever
+        if( g->m.ter( p ) == t_lava ) {
```

Comment:
`g->m.ter( p )` is invoked way too many times here. I'd extract it into a temporary variable. Would make conditions much shorter and easier to comprehend.